src/tsdb.c

```diff
@@ -279,72 +248,77 @@ void IndexMetricFromName(RedisModuleCtx *ctx, RedisModuleString *keyname) {
+        if (status) {
+            // Find the rule in the source key and rename the its destKey
+            CompactionRule *rule = srcSeries->rules;
+            while (rule) {
+                if (RedisModule_StringCompare(renameFromKey, rule->destKey) == 0) {
+                    RedisModule_FreeString(NULL, rule->destKey);
+                    RedisModule_RetainString(NULL, keyTo);
+                    rule->destKey = keyTo;
+                    break; // Only one src can point back to destKey
+                }
+                rule = rule->nextRule;
             }
-            rule = rule->nextRule;
+            RedisModule_CloseKey(srcKey);
         }
-        RedisModule_CloseKey(srcKey);
     }
 
     // A source key was renamed need to rename the srcKey on all the destKeys
-    if (series->rules) {
-        CompactionRule *rule = series->rules;
+    CompactionRule *rule = series->rules;
+    while (rule) {
         Series *destSeries;
         RedisModuleKey *destKey;
-        while (rule) {
-            const int status =
-                SilentGetSeries(ctx, rule->destKey, &destKey, &destSeries, REDISMODULE_WRITE);
-            if (!status) {
-                const char *destKeyName = RedisModule_StringPtrLen(rule->destKey, NULL);
-                RedisModule_Log(ctx,
-                                "warning",
-                                "couldn't open key or key is not a Timeseries. key=%s",
-                                destKeyName);
-            } else {
-                // rename the srcKey in the destKey
-                RedisModule_FreeString(NULL, destSeries->srcKey);
-                RedisModule_RetainString(NULL, keyTo);
-                destSeries->srcKey = keyTo;
+        CompactionRule *nextRule = rule->nextRule; // avoid iterator invalidation
+        const int status =
+            GetSeries(ctx, rule->destKey, &destKey, &destSeries, REDISMODULE_WRITE, false, false);
+        if (status) {
+            // rename the srcKey in the destKey
+            RedisModule_FreeString(NULL, destSeries->srcKey);
+            RedisModule_RetainString(NULL, keyTo);
+            destSeries->srcKey = keyTo;
 
-                RedisModule_CloseKey(destKey);
-            }
-            rule = rule->nextRule;
+            RedisModule_CloseKey(destKey);
         }
+        rule = nextRule;
+    }
+}
+
+void RenameSeriesTo(RedisModuleCtx *ctx, RedisModuleString *keyTo) {
+    // Try to open the series
+    Series *series;
+    RedisModuleKey *key = NULL;
+    const int status =
+        GetSeries(ctx, keyTo, &key, &series, REDISMODULE_READ | REDISMODULE_WRITE, true, true);
+    if (!status) { // Not a timeseries key
+        goto cleanup;
     }
 
+    // Reindex key by the new name
+    RemoveIndexedMetric(renameFromKey);
+    IndexMetric(keyTo, series->labels, series->labelsCount);
+
+    UpdateReferencesToRenamedSeries(ctx, series, keyTo);
```

Comment:
This is very clean and easy to read, good job on putting it in a separate function.